cpp/src/arrow/flight/sql/driver.cc

```diff
@@ -0,0 +1,1776 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <array>
+#include <cmath>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+#include "arrow/array/array_binary.h"
+#include "arrow/array/array_nested.h"
+#include "arrow/array/builder_base.h"
+#include "arrow/array/builder_binary.h"
+#include "arrow/array/builder_nested.h"
+#include "arrow/array/builder_primitive.h"
+#include "arrow/array/builder_union.h"
+#include "arrow/c/bridge.h"
+#include "arrow/config.h"
+#include "arrow/flight/client.h"
+#include "arrow/flight/sql/client.h"
+#include "arrow/flight/sql/driver_internal.h"
+#include "arrow/flight/sql/server.h"
+#include "arrow/flight/sql/types.h"
+#include "arrow/flight/sql/visibility.h"
+#include "arrow/io/memory.h"
+#include "arrow/io/type_fwd.h"
+#include "arrow/ipc/dictionary.h"
+#include "arrow/ipc/reader.h"
+#include "arrow/record_batch.h"
+#include "arrow/result.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/util/checked_cast.h"
+
+namespace arrow::flight::sql {
+
+using arrow::internal::checked_cast;
+
+namespace {
+/// \brief Client-side configuration to help paper over SQL dialect differences
+struct FlightSqlQuirks {
+  /// A mapping from Arrow type to SQL type string
```

Comment:
What is the current SQL dialect we are using here? maybe we should mention that in this comment.

Reply:
There's not a single SQL dialect here, this is here because we are trying to be a generic driver and we don't fully know the details of the server's dialect. So we need a bit of configuration to support some of the features specified in ADBC. Ideally we'd be able to infer these from the metadata Flight SQL provides. But if not (and because that's not implemented), there's some configuration escape hatches so you can make things work for you.

Reply:
Got it, thanks David!